site/en/blog/crx-scripting-api/index.md

```diff
@@ -0,0 +1,306 @@
+  let userReq = await fetch('/https://example.com/user-data.json');
+  let user = await userReq.json();
+  let givenName = user.givenName || '<GIVEN_NAME>';
+
+  chrome.scripting.executeScript({
+    target: { tabId: tab.id },
+    func: greetUser,
+    args: [givenName],
+  });
+});
+```
+
+### Targeting frames
+
+We also wanted to improve how developers interact with frames in the revised API. The Manifest V2
+version of `executeScript` allowed developers to target either all frames in a tab or a specific
+frame in the tab.
+
+```js
+// Manifest V2 extension
+chrome.browserAction.onClicked.addListener((tab) => {
+  chrome.webNavigation.getAllFrames({tabId: tab.id}, (frames) => {
+    let frame1 = frames[0].frameId;
+    let frame2 = frames[1].frameId;
+
+    chrome.tabs.executeScript(tab.id, {
+      frameId: frame1,
+      file: 'content-script.js',
+    });
+    chrome.tabs.executeScript(tab.id, {
+      frameId: frame2,
+      file: 'content-script.js',
+    });
+  });
+});
+```
+
+In Manifest V3, we replaced the optional `frameId` integer property in the options object with an
+optional `frameIds` array of integers; this allows developers to target multiple frames in a single
+API call. You can use use `chrome.webNavigation.getAllFrames` to get a list of all frames in a tab.
+
+```js
+// Manifest V3 extension
+chrome.action.onClicked.addListener(async (tab) => {
+  let frames = await new Promise((resolve) => {
+    chrome.webNavigation.getAllFrames({tabId: tab.id}, resolve);
+  });
+  let frame1 = frame[0].frameId;
+  let frame2 = frame[1].frameId;
+
+  chrome.scripting.executeScript({
+    target: {
+      tabId: tab.id,
+      frameIds: [frame1, frame2],
+    },
+    files: ['content-script.js'],
+  });
+});
+```
+
+### Script injection results
+
+We've also improved the way we return script injection results in Manifest V3. A "result" is
+basically the final statement evaluated in a script. Think of it like the value returned when you
+call `eval()` or execute a block of code in the Chrome DevTools console, but serialized in order to
+pass results across processes.
+
+In Manifest V2, `executeScript` and `insertCSS` would return an array of plain execution results.
+This is fine if you only have a single injection point, but result order is not guaranteed when
+injecting into multiple frames so there's no way to tell which result is associated with which
+frame.
+
+For a concrete example, let's take a look at the `results` arrays returned by a Manifest V2 and a
+Manifest V3 version of the same extension. Both versions of the extension will inject the same
+content script and we'll be comparing results on the same [demo page][iframe-demo].
+
+```js
+//// content-script.js
+var headers = document.querySelectorAll('h1,h2,h3,h4,h5,h6');
+headers.length;
+```
+
+When we run the Manifest V2 version, we get back an array of `[0, 1]`. Which result
+corresponds to the main frame and which is for the iframe? The return value doesn't tell us, so we
```

Comment:
Technically, first result is, in practice, guaranteed to be the main frame (even though we don't document that).  Can we use a demo with 3 frames?  If not, no big deal, because this is just an implementation detail that developers shouldn't be relying on.

Reply:
In the back of my head I thought that might be the case. I'm updating the return value example now.